automata/base/utils.py

```diff
@@ -1,12 +1,44 @@
 #!/usr/bin/env python3
 """Miscellaneous utility functions and classes."""
 
+import os
+import pathlib
+import random
+import uuid
 from collections import defaultdict
 from itertools import count, tee, zip_longest
-from typing import Any, Callable, Dict, Generic, Iterable, List, Set, Tuple, TypeVar
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Generic,
+    Iterable,
+    List,
+    Literal,
+    Set,
+    Tuple,
+    TypeAlias,
+    TypeVar,
+    Union,
+)
 
 from frozendict import frozendict
 
+# Optional imports for use with visual functionality
+try:
+    import pygraphviz as pgv
+except ImportError:
+    _visual_imports = False
+else:
+    _visual_imports = True
+finally:
+    # create a type for type checker
+    # irrespective of whether the imports were successful
+    GraphT: TypeAlias = "pgv.AGraph"
```

Comment:
I don't think that there is a need to use a string annotation here. I believe you can delete this block and add `from __future__ import annotations` at the top (as this will not read annotations unless code is being explicitly typechecked).